### PR TITLE
ReplicatedMergeTree - cleanup data directory after Zookeeper exceptions

### PR DESCRIPTION
**Pull request by bharatnc**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Cleanup data directory after Zookeeper exceptions during CreateQuery for StorageReplicatedMergeTree Engine.

Detailed description / Documentation draft:

It's possible for `getZooKeeper()` to timeout if  zookeeper host(s) can't
be reached. In such cases `Poco::Exception` is thrown after a connection
timeout - refer to `src/Common/ZooKeeper/ZooKeeperImpl.cpp:866` for more info.

Side effect of this is that the CreateQuery gets interrupted and it exits.
But the data Directories for the tables being created aren't cleaned up.
This unclean state will hinder table creation on any retries and will
complain that the Directory for table already exists.

To achieve a clean state on failed table creations, catch this error if
the exception is of type Poco::Exception and call `dropIfEmpty()` method,
then proceed throwing the exception. Without this, the Directory for the
tables need to be manually deleted before retrying the CreateQuery.

This addresses: #14517

### CODE REVIEW

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 1]
```diff
@@ -216,7 +216,30 @@ StorageReplicatedMergeTree::StorageReplicatedMergeTree(
         getStorageID().getFullTableName() + " (StorageReplicatedMergeTree::mutationsFinalizingTask)", [this] { mutationsFinalizingTask(); });
 
     if (global_context.hasZooKeeper())
-        current_zookeeper = global_context.getZooKeeper();
+    {
```
Comment (bharatnc) [inline src/Storages/StorageReplicatedMergeTree.cpp:219]: I have tested this locally,  and works great, but would like to add an integration / unit test? Still thinking of the best/right way to do that? Would be useful to have suggestions. Simplest test steps wil be:

1. Setup bad zookeeper hosts in the config file.
2. Create database and then try creating the ReplicatedMergeTreeTable.
3. Expect exception.
4. Try recreating the same table.
5. Should not throw error that the Data directory already exists for the table.

Comment (bharatnc) [reply src/Storages/StorageReplicatedMergeTree.cpp:219]: cc: @filimonov @alexey-milovidov

Comment (bharatnc) [reply src/Storages/StorageReplicatedMergeTree.cpp:219]: Added integration tests that test the behavior and verified that this test works fine locally.

**Comment by bharatnc** [comment]
Also addresses  #14115

**Review by tavplubix — APPROVED** [review]

**Comment by bharatnc** [comment]
@tavplubix  thank you for adding those tests. Should this PR be considered a `Bug`. I had initially created it under `Improvement` - but makes more sense to be a bug. I did update the PR description, but label needs to be changed. Do let me know if this should rather be an`improvement`.

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 2]
```diff
@@ -216,7 +216,30 @@ StorageReplicatedMergeTree::StorageReplicatedMergeTree(
...
+        /// the excaption is of type Poco::Exception and call dropIfEmpty() method,
+        /// then proceed throwing the exception. Without this, the Directory for the
+        /// tables need to be manually deleted before retrying the CreateQuery.
+        try
+        {
+            current_zookeeper = global_context.getZooKeeper();
+        }
+        catch (Poco::Exception & e)
+        {
+            dropIfEmpty();
```
Comment (filimonov) [inline src/Storages/StorageReplicatedMergeTree.cpp:239]: So folders are created in MergeTreeData constructor?

I don't think we should do any destructive operations (even removing empty folders) during ATTACH. 
```cpp
 if (!attach) .... 
```

it seem you  don't need to capture the exception in the variable, so you can do this way:

https://github.com/ClickHouse/ClickHouse/blob/dee1fefeb472bebdfa4e15eb1b7ec7e428069f8a/src/Storages/StorageReplicatedMergeTree.cpp#L321-L325

Comment (tavplubix) [reply src/Storages/StorageReplicatedMergeTree.cpp]: Thanks for suugestions

**Comment by tavplubix** [comment]
> @tavplubix thank you for adding those tests. Should this PR be considered a `Bug`. I had initially created it under `Improvement` - but makes more sense to be a bug. I did update the PR description, but label needs to be changed. Do let me know if this should rather be an`Improvement`.

Ok, let's change the changelog category to "Bug Fix", so it will be backported

File: src/Storages/StorageReplicatedMergeTree.cpp (hunk 2 above)
Comment (bharatnc) [reply src/Storages/StorageReplicatedMergeTree.cpp]: Thanks you. Also just made minor changes to the comment to reflect that this is what we're doing.

**Pull request merged**